worlds/witness/test/__init__.py

```diff
@@ -0,0 +1,86 @@
+from test.bases import WorldTestBase
+from test.general import gen_steps, setup_multiworld
+from test.multiworld.test_multiworlds import MultiworldTestBase
+from typing import Any, ClassVar, Dict, Iterable, List, Mapping, Union
+
+from BaseClasses import CollectionState, Item
+
+from worlds.witness import WitnessWorld
+
+
+class WitnessTestBase(WorldTestBase):
+    game = "The Witness"
+    player: ClassVar[int] = 1
+
+    world: WitnessWorld
+
+    def can_beat_game_with_items(self, items: Iterable[Item]) -> bool:
+        state = CollectionState(self.multiworld)
+        for item in items:
+            state.collect(item)
+        return state.multiworld.can_beat_game(state)
+
+    def assert_can_beat_with_minimally(self, required_item_counts: Mapping[str, int]):
+        """
+        Assert that the specified mapping of items is enough to beat the game,
+        and that having one less of any item would result in the game being unbeatable.
+        """
+        # Find the actual items
+        found_items = [item for item in self.multiworld.get_items() if item.name in required_item_counts]
+        actual_items = {item_name: [] for item_name in required_item_counts}
+        for item in found_items:
+            if len(actual_items[item.name]) < required_item_counts[item.name]:
+                actual_items[item.name].append(item)
+
+        # Assert that enough items exist in the item pool to satisfy the specified required counts
+        for item_name, item_objects in actual_items.items():
+            self.assertEqual(
+                len(item_objects),
+                required_item_counts[item_name],
+                f"Couldn't find {required_item_counts[item_name]} copies of item {item_name} available in the pool, "
+                f"only found {len(item_objects)}",
+            )
+
+        # assert that multiworld is beatable with the items specified
+        self.assertTrue(
+            self.can_beat_game_with_items(item for items in actual_items.values() for item in items),
+            f"Could not beat game with items: {required_item_counts}",
+        )
+
+        # assert that one less copy of any item would result in the multiworld being unbeatable
+        for item_name, item_objects in actual_items.items():
```

Comment:
I would personally subtest this one, so that you know more clearly which items are failing, and most importantly, if more than one are failing, you don't need to run the test many times to find them all

Reply:
Hmmm, how do you mean? I'm not super experienced with unit tests - Since this is a custom *assert* function and not a custom test, can I still have a subtest within it?

Reply:
I might be misunderstanding how these custom asserts work, but here's what I have in mind. You can tell me if it doesn't work, or if you're against it for any reason
```py
for item_name, item_objects in actual_items.items():
    with self.subtest(f"Verify cannot beat world without {item_name}"):
        removed_item = item_objects.pop()
        self.assertFalse(
            self.can_beat_game_with_items(item for items in actual_items.values() for item in items),
            f"Game was beatable despite having {len(item_objects)} copies of {item_name} "
            f"instead of the specified {required_item_counts[item_name]}",
        )
        item_objects.append(removed_item)
```

Reply:
Ohhh I'll try this, sure

Reply:
Unfortunately, I still seem to only get one output when there are two wrong items

Reply:
What test runner are you using? I know I didn't get subtests working properly when I was using pytest

Reply:
Oh yeah it is indeed pytest

Reply:
What do you use now?

Reply:
I run my tests with unittest nowadays, and subtests work fine. I'd give it a shot, just to know if the syntax works. Then, at least the test will exist, if it does work

Reply:
Using subtest right now. `unittest` is a bit annoying because it seems like it's not possible to do `unittest worlds/witness` the same way I can `pytest worlds/witness`. But I did confirm that it works with `unittest` if I just run all tests, I am getting two outputs. And it's not *broken* with pytest, it just doesn't give me two outputs. But that's ok